### ISSUE TITLE
Question about SPA/Implicit type workflow

### ISSUE THREAD

**Issue by dmitrizagidulin** [bspk/oauth.xyz-site#8]
I'm having trouble picturing the work flow that an in-browser JS SPA type client would use, with xyz. Can you help walk me through this?

So, say my goal is to get some user credential, like an ID token. I have a Login button. User clicks, it opens a popup window (that's going to do the browser redirects, so as not to mess with the main application - a common requirement). 
Let's skip the webfinger / provider selection step - let's say the AS is hardcoded.

I assume the client (inside the popup) makes a POST to the /transaction endpoint:

```js
{
    "client_display": {
        "name": "My Client Display Name",
        "uri": "https://example.net/client"
    },
    "resources": [
        "id_token" // not sure how to indicate that an ID token is wanted
    ],
    "interact": {
        "type": "redirect",
        "callback": "https://client.example.net/callback"
    },
    key: { /* ... */ }
}
```

Gets back a server response:

```js
{
    "interaction_url": "https://server.example.com/interact/4CF492MLVMSW9MKMXKHQ",
    "server_nonce": "MBDOFXG4Y5CVJCX821LH",
     "handle": {
        "value": "80UPRY5NM33OMUKMKSKU",
        "type": "bearer"
    },
    /* various other handles */
}
```
At this point, the client (inside the popup window) is supposed to redirect that window to the `interaction_url`, yeah? (And then I assume the usual login & consent step takes place).
Once the user authenticates to the AS/IdP, they're supposed to redirect the user back to the `interact.callback` uri?

So assuming I'm on the right track, my question is this:

Once the AS redirects the user back to the client, how does the client actually get the desired credential? Do they make a continuation request with the `handle` they received in the initial response?

**Comment by jricher** [comment]
Yes, that's exactly right -- they make a continuation request with the transaction handle from the first response and the interaction handle from the front-channel response, together.

**Comment by jricher** [comment]
For example:

```
{
    "handle": "80UPRY5NM33OMUKMKSKU",
    "interact_handle": "CuD9MrpSXVKvvI6dN1awtNLx-HhZy46hJFDBicG4KoZaCmBofvqPxtm7CDMTsUFuvcmLwi_zUN70cCvalI6ENw"
}
```

**Comment by dmitrizagidulin** [comment]
@jricher - is the `interact_handle` even needed, for that continuation request? (As in, can't the server keep track of state just via the TX handle?

**Comment by jricher** [comment]
The `interact_handle` is to prove that the client is responding to a specific front-channel incoming request associated with the transaction. It closes the loop more securely than just polling, which is susceptible to a fixation attack. This is equivalent to the authorization code in OAuth2.

**Comment by dmitrizagidulin** [comment]
@jricher Interesting. Makes sense I think. (Although I'm not sure why the client would need to poll - doesn't the ending of their front-end interaction signal that they can go back and send the request again, just using `handle`?)

**Comment by jricher** [comment]
The client wouldn't poll if it can listen for an inbound front-channel response, generally speaking. But there are many clients that can't accept inbound connections/signals, and so they can poll.

**Comment by dmitrizagidulin** [comment]
Wait, so there's a piece I'm clearly missing. What do you mean by an inbound front-channel response?

Is the idea (given our setup here, an in-browser JS app with a popup window for user sign in) that once the user finishes authentication, and the AS redirects them back to the `interact.callback.uri`, it passes along the `interact_handle` param?

So that, say once the user submits the login form (at the AS), and gets a 302 redirect back to:

`https://client.example.net/callback#interact_handle=CuD9MrpSXVKvvI6dN1awtNLx-...`  

Is that the idea?

**Comment by jricher** [comment]
Inbound front-channel response is the redirect back to the client application. So you're Correct in your example, but for a minor detail as the value would be as a query parameter and not a fragment. It's no different from a webserver or native app flow, since in all of these cases you can get the user out to an arbitrary URL and have a way of listening for a return from the browser on a callback URL. What you need now is the client to complete that connection when it continues the transaction, to prove that it's the same piece of software that had started the transaction.

**Comment by dmitrizagidulin** [comment]
>  the value would be as a query parameter and not a fragment.

Right, so, I noticed that, that all the examples used query params exclusively. Except that hash fragment params also need to be supported (since they don't leak their values to the host's server / end up in access logs).

**Comment by jricher** [comment]
> since they don't leak

That's not universally true, and there are numerous problems with using them as they get stuck across redirects and pinned into browser history in ways people don't expect. It's caused significant issues in OAuth and so I'm avoiding that by going only with query params. Also the front channel doesn't return anything that's usable on its own anyway, so returning those parameters to the server doesn't leak anything.

**Comment by dmitrizagidulin** [comment]
@jricher hmm, but I thought _both_ query params and hash params have caused problems in OAuth, historically. And that query params caused _more_. 

Are you sure it make sense to just pick query param exclusively? Like, I get that you want to simplify the design decisions implementers have to make. But it's not an inconsequential tradeoff.

(And also, would you entertain the idea of supporting the form_post return type? That's neither query or hash fragment params, for apps that can support it?)

**Comment by jricher** [comment]
Query params have not caused more, in my experience. The flexibility of different types doesn't buy anything that I can see, and it's caused a ton of problems for developers (especially client developers) in going through all of the permutations available in OIDC, for instance. 

Additionally, form post is really only beneficial when you've got a large payload, which is something I explicitly want to avoid here. I want to get as much OUT of the front channel as possible, so I see these two return parameters as the minimum needed for the protocol to function.

**Comment by dmitrizagidulin** [comment]
K. Clear, thanks!

**Comment by dmitrizagidulin** [comment]
(Gonna leave this issue open as a reminder to myself to extract this discussion into an example on the site)
